src/cljs/nr/utils.cljs

```diff
@@ -91,3 +92,138 @@
+          "[hb]" "haas-bioroid"
+          "[haas-bioroid]" "haas-bioroid"
+          "[jinteki]" "jinteki"
+          "[nbn]" "nbn"
+          "[shaper]" "shaper"
+          "[sunny]" "sunny"
+          "[weyland]" "weyland-consortium"
+          "[weyland-consortium]" "weyland-consortium"}
+      (map (fn [[k v]] [(regex-of k) (span-of v)])))))
+
+(defn card-smap-impl []
+  "A map of case sensitive card regexes to card span fragments"
+  (letfn [(unpack-card [[title cards]] [title (:code (first cards))])
+          (span-of [title code] [:span {:class "fake-link" :id code} title])
+          (regex-of [card-title] (re-pattern (regex-escape card-title)))]
+    (->> @all-cards
+         (filter #(not (:replaced_by %)))
+         (group-by :title)
+         (map unpack-card)
+         (map (fn [[k v]] [(regex-of k) (span-of k v)]))
+         (into {}))))
+
+(def card-smap (memoize card-smap-impl))
+
+(defn ordered-keys-impl [smap]
+  "List the keys of a hashmap by length after stringifying them"
+  (sort-by (comp count str first) > smap))
+
+(def ordered-keys (memoize ordered-keys-impl))
+
+(defn padded-interleave [pad & seqs]
+  "Interleave sequences of uneven lengths by padding out the shorter ones"
+  (let [lazy-padded-seqs (map #(concat % (repeat pad)) seqs)
+        num-seqs (count lazy-padded-seqs)
+        max-len (reduce max (map count seqs))]
+    (take (* max-len num-seqs) (apply interleave lazy-padded-seqs))))
+
+(defn replace-in-element [element [pattern replacement]]
+  "Given a string element, split that string on pattern boundaries and replace
+  all patterns with a provided replacement. If element is not a string, return
+  it unmodified."
+  (if (string? element)
+    (let [context (split element pattern)
+          match-count (count (re-seq pattern element))
+          replacements (repeat match-count replacement)]
+      (->> (padded-interleave "" context replacements)
+           (filter not-empty)))
+    [element]))
+
+(defn replace-in-fragment [fragment substitution]
+  "Split all string elements in a fragment and replace the splitting element
+  according to substitution"
+  (reduce concat (map #(replace-in-element % substitution) fragment)))
+
+(defn set-react-key [n elem]
+  "Given a reagent-style HTML element, set the :key attribute of the element"
+  (let [head (first elem)
+        attr (if (map? (second elem)) (second elem) {})
+        tail (if (map? (second elem)) (drop 2 elem) (drop 1 elem))]
+  (into [] (concat [head (merge attr {:key n})] tail))))
+
+(defn render-fragment-impl [fragment replacement-smap]
+  "Given a fragment, shallowly replaces text in the fragment with icon and,
+  optionally, card preview HTML"
+  (let [counter (atom 0)
+        set-next-key (fn [elem] (set-react-key (do (swap! counter inc) @counter) elem))]
+    (->> (reduce replace-in-fragment fragment (ordered-keys replacement-smap))
+         (replace replacement-smap)
+         (map-if vector? set-next-key)
```

Comment:
I think `map-if` should either be inlined or defined in the above let-binding. Having it at the top of the file makes discoverability and understanding harder.

Reply:
Sounds good to me :+1: